systemd/ignition-firstboot-complete.service

```diff
@@ -18,7 +18,12 @@ RemainAfterExit=yes
 # detected this file. Fail if we are unable to remove it, rather than risking
 # rerunning Ignition at next boot.
 MountFlags=slave
-ExecStart=/bin/sh -c 'mount -o remount,rw /boot && rm /boot/ignition.firstboot'
+# It is better to have a separate script to do this but it might be polutting
+# the target system with some script in i.e. /usr/sbin/firstboot-complete
+# The retval code is still respected with having this if-else block.
+ExecStart=/bin/sh -c \
+	'mount -o remount,rw /boot && rm /boot/ignition.firstboot && \
+	if [[ $(uname -m) = s390x ]]; then zipl; fi'
```

Comment:
We could ship this and be totally fine but...the perfectionist in me is bothered by the potential case here where the system is interrupted between `rm /boot/ignition.firstboot` and running `zipl` - we'd be in a situation where Ignition would rerun every boot until an upgrade (or something else ran `zipl`).  (Actually wait, have we even patched libostree to run `zipl`?)
Anyways so I think we shoudl remove the stamp file *after* running zipl, i.e.:

```
ExecStart=/bin/sh -c 'mount -o remount,rw /boot && if [[ $(uname -m) = s390x ]]; then zipl; fi && rm /boot/ignition.firstboot '
```
or so?

Reply:
`(Actually wait, have we even patched libostree to run zipl?)` : The only attempt I made to patch libostree for zipl support is to generate the config file - similar to this https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-bootloader-syslinux.c for ` boot/syslinux/syslinux.cfg` - which I think we don't need. Because zipl understands BLS and we would also ship a minimal zipl.conf (soon to be merged and backported). For other means outside generating the config, I'm sure if we need to patch libostree for zipl. Please some light on it (either here or on IRC).
And agree with you about moving the `rm` after the `if` block, didn't really consider that details. Good point

Reply:
zipl understands BLS, but the command still needs to be rerun after updating the BLS files, right?  (Unlike grub2)

Reply:
And if I'm not wrong, s390x/zipl doesn't care if `/boot/ignition.firstboot` exists or not - at least for now when we don't have grub-emu yet (as described in#84). AFAIU, `/boot/ignition.firstboot` only makes sense for grub.

Reply:
`zipl understands BLS, but the command still needs to be rerun after updating the BLS files, right? (Unlike grub2)` : Ah right, now I remember we once talked about the hook/trigger mechanism to invoke zipl after ostree updates new commits.

Reply:
Is this a matter of adding a `ostree-bootloader-zipl.c` into `ostree` or can we utilize a simple hook for `rpm-ostree`?

Reply:
when I simulate an rpmostree upgrade by creating a local branch and rebasing to that branch i get 

```
[  178.292569] ostree-prepare-root[2547]: ostree-prepare-root: Couldn't find specified OSTree root '/sysroot//ostree/boot.0/rhcos/431aee16642ff2306bb5f5f0d72b5b57c20eb11964ebfc86755abb19768977d5/0': No such file or directory
```
is the upgrade error related to this issue of re running zipl ?

Reply:
Could you provide more details ?

Reply:
I did a coreos build of `rhcos-4.2`using the coreos-assembler `4.2-multiarch` branch. I executed a  "cosa run" to get into the VM and executed these commands to upgrade (which is what the kola rpmostree upgrade test does):

```
sudo ostree refs --create local-branch c71f7d816bee20ab1248afc73e78149cae9771ca550153529e59b1a4003c2b3b
sudo ostree commit -b local-branch --tree ref=c71f7d816bee20ab1248afc73e78149cae9771ca550153529e59b1a4003c2b3b --add-metadata-string version=kola-test-1.0
sudo rpm-ostree rebase : local-branch
```

After a reboot ,  I see the above error message repeated over and over again. That being said I haven't tried with FCOS. Maybe I should give that a try

Reply:
I think this commit (with changes when backported, if any) will partially fix the issue you are having.
The reason is, zipl was not re-run to apply new ostree commit in new BLS file, thus it still looks for old BLS entry.
I said partially because there are 3 other changes you might need (which I'm also testing at the same time) : s390-tools/pull/71, s390-tools/pull/74, coreos-assembler/pull/780.

Reply:
I believe that error is happening exactly because libostree isn't running `zipl` - if the bootloader configuration is stale you could see this.

Reply:
To say this clearly: Until we have the libostree `zipl` backend, upgrades will not work.

Reply:
That's correct.

Reply:
Ok...thank you!

Reply:
> To say this clearly: Until we have the libostree zipl backend, upgrades will not work.

@tuan-hoang1 @Prashanth684 are you folks working on a `zipl` backend?

Reply:
@tuan-hoang1 have you started on this ^ . If not I can help with the libostree backend..please let me know...

Reply:
Let's work on this in https://github.com/ostreedev/ostree/pull/1937 - I can add you guys as collaborators for libostree so you can push to the PR.  Or fork the code in that PR and push a new one that you have access to (and feel free to take ownership and use [co-authored-by](https://help.github.com/en/articles/creating-a-commit-with-multiple-authors) with me).

Reply:
Couldn't be better, working on it, quite urgent tbh. Many thanks, Colin.

Reply:
Ok @cgwalters @tuan-hoang1 the zipl backend works well on rhcos with cosa master. The other question i had is - would this work for rhcos-4.2 with cosa-4.2 (basically using anaconda) and how do we deal with setting the backend to zipl like we do in create_disk.sh?

Reply:
@Prashanth684 I think of 2 ways:
1. Add `ostree config set bootloader` to anaconda patch at https://bugzilla.redhat.com/show_bug.cgi?id=1741908  and generate new `updates.img`.
2. Execute `ostree config set bootloader` in `gf-anaconda-cleanup`.

Reply:
But that's another issue. Let's not hijack this PR :P

Reply:
> Execute ostree config set bootloader in gf-anaconda-cleanup

I like doing this fix in cosa - changing Anaconda is much harder.  It might also work to do it in a `%post --nochroot` in `image-base.ks`.